runtime/lua/vim/fs.lua

```diff
@@ -334,6 +327,46 @@ function M.find(names, opts)
   return matches
 end
 
+--- Find the first parent directory containing a specific "marker", relative to a buffer's
+--- directory.
+---
+--- Example:
+---
+--- ```lua
+--- -- Find the root of a Python project
+--- vim.fs.findroot(0, {'pyproject.toml', 'setup.py' })
+---
+--- -- Find the root of a git repository
+--- vim.fs.findroot(0, '.git')
+---
+--- -- Find the parent directory containing any file with a .csproj extension
+--- vim.fs.findroot(0, function(name, path)
+---   return name:match('%.csproj$') ~= nil
+--- end)
+--- ```
+---
+--- @param bufnr integer Buffer number, or 0 for current buffer.
```

Comment:
So far `fs.lua` has been completely agnostic to vim buffers, and only operates on strings paths. Do we really want to break that pattern now?

Reply:
Would work for the lsp use case just as well (with `args.file` -- but see above re: expanding). Given the choice, I too would prefer filename over buffer name (which might not exist yet...)

Reply:
True, but `vim.fs` intentionally also has "path" concepts to avoid too much dogma. After all, this is still a text editor, so the modules are intended to be "useful" rather than 100% pure.

> this function should accept a string path argument to make it more general.

I thought about that too, it's not a one-way door though?

Reply:
> So far `fs.lua` has been completely agnostic to vim buffers, and only operates on strings paths. Do we really want to break that pattern now?

I had the same realization. There's not a better place to put it (for now) though. If we already had `vim.projects` or similar this would be a good fit for that, but since we don't, `vim.fs` is the next best spot.

> And either way, this function should accept a string path argument to make it more general.

I'll gladly make it a string _or_ buffer number, but don't want to make it just a string, since the entire _raison d'être_ of this function is to eliminate boilerplate, and `vim.fs.findroot(vim.api.nvim_buf_get_name(0), ...` does not serve that objective.

Reply:
8dca2167f5f6fe578a15211aa6cf4fb76b1c8d32

Reply:
>Where is a named buffer (rather than a file) strictly needed?

ftplugin